addons/addon-base-raas/packages/base-raas-services/lib/environment/environment-mount-service.js

```diff
@@ -212,15 +237,463 @@ class EnvironmentMountService extends Service {
+   * for access to the study
+   *
+   * @param {Object} updateRequest - permission add/remove/update requests coming from SWB UI
+   * @returns {Object[]} - List of users which newly/continue-to have access to the study
+   */
+  _getAllowedUsers(updateRequest) {
+    return _.filter(
+      updateRequest.usersToAdd,
+      userToAdd =>
+        !_.includes(
+          _.map(
+            _.filter(updateRequest.usersToRemove, u => u.permissionLevel !== 'admin'),
+            userToRemove => userToRemove.uid,
+          ),
+          userToAdd.uid,
+        ) && userToAdd.permissionLevel !== 'admin',
+    );
+  }
+
+  /**
+   * This method will filter out users from the original updateRequest
+   * who lost access to the study
+   *
+   * @param {Object} updateRequest - permission add/remove/update requests coming from SWB UI
+   * @returns {Object[]} - List of users which lost access to the study
+   */
+  _getDisAllowedUsers(updateRequest) {
+    return _.filter(
+      updateRequest.usersToRemove,
+      userToRemove =>
+        !_.includes(
+          _.map(
+            _.filter(updateRequest.usersToAdd, u => u.permissionLevel !== 'admin'),
+            userToAdd => userToAdd.uid,
+          ),
+          userToRemove.uid,
+        ) && userToRemove.permissionLevel !== 'admin',
+    );
+  }
+
+  /**
+   * This method will filter out users from the original updateRequest
+   * who have switched permission levels for the study
+   *
+   * @param {Object} updateRequest - permission add/remove/update requests coming from SWB UI
+   * @returns {Object[]} - List of users which have switched permission levels for the study
+   */
+  _getPermissionChangeUsers(updateRequest) {
+    return _.filter(
+      updateRequest.usersToRemove,
+      userToRemove =>
+        _.includes(
+          _.map(
+            _.filter(updateRequest.usersToAdd, u => u.permissionLevel !== 'admin'),
+            userToAdd => userToAdd.uid,
+          ),
+          userToRemove.uid,
+        ) && userToRemove.permissionLevel !== 'admin',
+    );
+  }
+
+  /**
+   * Function that returns updated policy document after ensuring workspace has list permissions for studies it has some access to
+   *
+   * @param {Object} policyDoc - S3 studydata policy document for workspace role
+   * @param {String} studyPathArn
+   * @returns {Object[]} - the statement to update in the policy
+   */
+  _ensureListAccess(policyDoc, studyPathArn) {
+    const s3BucketName = studyPathArn.substring(0, studyPathArn.indexOf('/'));
+    const studyPrefix = studyPathArn.substring(studyPathArn.indexOf('/') + 1);
+
+    _.forEach(policyDoc.Statement, statement => {
+      if (
+        statement.Sid.startsWith('studyListS3Access') &&
+        statement.Effect === 'Allow' &&
+        statement.Resource === s3BucketName &&
+        !_.includes(statement.Condition.StringLike['s3:prefix'], studyPrefix)
+      ) {
+        statement.Condition.StringLike['s3:prefix'].push(studyPrefix);
+      }
```

Comment:
Shouldn't the if condition be evaluated after iterating through all the statements and then if all the statements didn't include the studyPrefix, only then we should push the studyPrefix.

Reply:
There is only one studyListS3Access statement per bucket which holds all the studyPrefixes for that bucket, therefore only one in that list that would need that change. 

Even if access to all the studies in a certain bucket are taken away access from this policy, the studyListS3Access statement would still be present for this bucket with Condition.StringLike being an empty list.

Reply:
Got it, thanks.